nemo_aligner/models/nlp/gpt/megatron_gpt_spin_model.py

```diff
@@ -0,0 +1,524 @@
+        metrics = {"loss": loss_value}
+
+        return loss_value, metrics
+
+    def prepare_for_training_step(self):
+        # custom trainers will always zero grad for us
+        prepare_for_training_step(self, zero_grad=False)
+
+    def prepare_for_training(self):
+        configure_batch_sizes(
+            mbs=self.cfg.micro_batch_size,
+            gbs=self.cfg.global_batch_size,
+            dp=parallel_state.get_data_parallel_world_size(),
+        )
+        self.onload_adam_states()
+
+    def finish_training_step(self):
+        grad_reductions(self)
+
+    def finish_training(self):
+        """no need to offload adam states here
+        """
+
+    def prepare_for_validation(self):
+        configure_batch_sizes(
+            mbs=self.cfg.micro_batch_size,
+            gbs=self.cfg.global_batch_size,
+            dp=parallel_state.get_data_parallel_world_size(),
+        )
+
+    def prepare_for_validation_step(self):
+        prepare_for_validation_step(self)
+
+    def finish_validation_step(self):
+        finish_validation_step(self)
+
+    def finish_validation(self):
+        """no need to offload adam states here
+        """
+
+    def prepare_for_inference(self):
+        """normally we would configure the micro batch calculator here
+            but the nemo generation already does the configuration"""
+        self._reset_activation_checkpointing_args()
+        self._reset_sequence_parallelism_args()
+        set_eval(self)
+        self.offload_adam_states()
+
+    def finish_inference(self):
+        # training will onload the adam states, no need to onload it here
+        self._restore_activation_checkpointing_args()
+        self._restore_sequence_parallelism_args()
+        set_train(self)
+
+    def offload_adam_states(self):
+        if self.distributed_adam_offload_manager is None:
+
+            self.distributed_adam_offload_manager = (
+                offload_distributed_adam(self._optimizer.state_dict(state_dict_format=1, gather_on_root=False))
+                if self.to_offload_adam_states and self.with_distributed_adam
+                else nullcontext()
+            )
+
+            # offload onto cpu
+            self.distributed_adam_offload_manager.__enter__()
+
+    def onload_adam_states(self):
+        if self.distributed_adam_offload_manager is not None:
+            # load back onto GPU
+            self.distributed_adam_offload_manager.__exit__(None, None, None)
+
+        self.distributed_adam_offload_manager = None
+
+    def sharded_state_dict(self, prefix: str = ""):
+        sharded_state_dict = super().sharded_state_dict(prefix=prefix)
+
+        # add in the reference policy weights
+        if self.ref_policy_state_dict is not None:
+            sharded_state_dict["reference_policy"] = make_sharded_tensors_for_checkpoint(
+                self.ref_policy_state_dict, prefix
+            )
+
+        return sharded_state_dict
+
+    def on_load_checkpoint(self, checkpoint) -> None:
```

Comment:
I guess there are still some issues with re-loading the reference policy's state dict based on what you said earlier this week. At first sight I'm not sure where the issue is, but I'm not familiar enough with this distributed checkpointing stuff.

One thing I was wondering is you may remember our discussion where I mentioned https://github.com/NVIDIA/NeMo/blob/bb575b72fd0be51ae10cc77d9f89ddb9e9d3b96d/nemo/collections/nlp/parts/nlp_overrides.py#L338 which erases `state_dict`. So I'm curious why you would still use it here... (but again, I really don't have a full understanding of the code flow).